Simplify resource-related variables, remove CDUMP where unneeded

DavidHuber-NOAA opened the pull request:
# Description
This overhauls resource-related variables to use a common set of variables for each job.  In the process, this also removed the use of CDUMP in most cases.

Resolves #1299 #2693

Depends on https://github.com/NOAA-EMC/wxflow/pull/35

# Type of change
- New feature (adds functionality)
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Created .xml and compared against develop.  Other tests to follow on Hera and WCOSS2:
- [x] C48_ATM
- [x] C48_S2SW
- [x] C48_S2SWA_gefs
- [x] C48mx500_3DVarAOWCDA (hera only)
- [x] C96C48_hybatmDA
- [x] C96C48_ufs_hybatmDA
- [x] C96_atm3DVar_extended (wcoss2 only)
- [x] C96_atmaerosnowDA

And basic tests on Hercules:
- [x] C48_ATM
- [x] C48_S2SW
- [x] C96C48_hybatmDA

# Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

DavidHuber-NOAA committed:
Simplify resource declarations in config.resources

(cherry picked from commit 130e2e479fbb625cc5b57d39aa0802c711e2916c)

DavidHuber-NOAA committed:
Remove 'step' from resource variable names in env files

(cherry picked from commit c3276ac77b2c5411a0e8b769e8957041f71e62e8)

DavidHuber-NOAA committed:
Trim env files by declaring common variables

(cherry picked from commit 325891d9bed5482017cebff52bb1e3c3d15eaf40)

DavidHuber-NOAA committed:
More env cleanup

DavidHuber-NOAA committed:
Get sfcanl npe based on ntiles

DavidHuber-NOAA committed:
Export standard resource variables only

DavidHuber-NOAA committed:
Clean up machine resource files

DavidHuber-NOAA committed:
Improved variable name

DavidHuber-NOAA committed:
Strip cdump from resource variables in tasks.py

DavidHuber-NOAA committed:
Replace references of CDUMP with RUN

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Activate snow DA test on WCOSS (#2720)
  Cleanup of stale RUNDIRS from an experiment (#2719)
  Update logic for MOM6 number of layers/exception values (#2681)
  Update wave jobs to use COMIN/COMOUT (#2643)

DavidHuber-NOAA committed:
Add whitespace

DavidHuber-NOAA committed:
Add call to update_base

DavidHuber-NOAA committed:
Update wxflow to enable setting envars before sourcing configs

DavidHuber-NOAA committed:
Update wxflow to include new test

DavidHuber-NOAA committed:
Revert wxflow

DavidHuber-NOAA committed:
Create temporary 'config.run's to set RUN at setup time

DavidHuber-NOAA committed:
Replace (()) with {}

DavidHuber-NOAA committed:
Ignore 'misspellings'

DavidHuber-NOAA committed:
Ignore 'unused' variables

DavidHuber-NOAA committed:
Address shellcheck errors

DavidHuber-NOAA committed:
Revert GAEA.env, remove unused APRUN

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  updated Finalize in Jenkinsfile and added try block around scm checkout (#2692)

DavidHuber-NOAA committed:
Update gefs resources to clean resource variables

DavidHuber-NOAA committed:
Redefine npe in analysis jobs

DavidHuber-NOAA committed:
Remove more references to CDUMP

DavidHuber-NOAA committed:
Update wxflow so CDUMP isn't required

DavidHuber-NOAA committed:
Add CDUMP cludge until wxflow is updated

DavidHuber-NOAA committed:
Separate wtimes for eobs and eomg

DavidHuber-NOAA committed:
Revert wxflow

DavidHuber-NOAA committed:
Check stage RUN instead of CDUMP

DavidHuber-NOAA committed:
Rename update_base to match superclass

DavidHuber-NOAA committed:
Make a copy of appconfig to for each RUN as it is modified

DavidHuber-NOAA committed:
Add envvar notes to create_experiment.py

DavidHuber-NOAA committed:
Check that common variables are defined before using them

DavidHuber-NOAA committed:
Switch analcalc CDUMP to RUN

DavidHuber-NOAA committed:
Fix postsndcfp resource declaration

DavidHuber-NOAA committed:
Further simplify analysis resources

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA committed:
Ignore 'unused' SC warning

DavidHuber-NOAA committed:
Replace CDUMP with RUN in docs

DavidHuber-NOAA committed:
Remove unused CDUMP references

DavidHuber-NOAA committed:
Revert CDUMP in analysis scripts to strip 'enkf'

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA committed:
Export RUN before sourcing configs

DavidHuber-NOAA committed:
Update GEFS resources after last merge

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA commented:
Marking ready for review after successful completion of multiple tests on Hera and WCOSS2.  However, this PR depends on https://github.com/NOAA-EMC/wxflow/pull/35, which needs to be merged first.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into HEAD

* emc/develop:
  Update gempak job to run one fcst hour per task (#2671)

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Hotfix for undefined CLUSTERS (#2748)

DavidHuber-NOAA committed:
Add todo to remove CDUMP from the ocean_analysis_post j-job

DavidHuber-NOAA committed:
Rename CDUMP to rCDUMP for analysis jobs that need exporting

DavidHuber-NOAA committed:
Remove CDUMP from ecf files; do not export CDUMP in J-Jobs

DavidHuber-NOAA committed:
Restore CDUMP to Fit2Obs J-job

DavidHuber-NOAA committed:
Restore CDUMP to Fit2Obs (part 2).

DavidHuber-NOAA commented on parm/config/gfs/config.resources:
```diff
@@ -126,149 +127,135 @@ export npe_node_max
+    wtime="00:10:00"
+    npe_gdas=5
+    npe_gfs=65
+    nth_gfs=1
+    npe_node_gdas=$(( npe_node_max / nth_gdas ))
+    npe_node_gfs=$(( npe_node_max / nth_gfs ))
```
```suggestion
    nth=1

    npe_node=$(( npe_node_max / nth ))
```

DavidHuber-NOAA committed:
Update parm/config/gfs/config.resources

DavidHuber-NOAA commented on parm/config/gfs/config.resources:
```diff
@@ -440,47 +408,38 @@ case ${step} in
-    export nth_aeroanlrun_gfs=1
-    export npe_node_aeroanlrun_gdas=$(( npe_node_max / nth_aeroanlrun_gdas ))
-    export npe_node_aeroanlrun_gfs=$(( npe_node_max / nth_aeroanlrun_gfs ))
+    wtime="00:30:00"
+    npe_gdas=$(( layout_x * layout_y * 6 ))
+    npe_gfs=$(( layout_x * layout_y * 6 ))
+    nth_gdas=1
+    nth_gfs=1
+    npe_node_gdas=$(( npe_node_max / nth_gdas ))
+    npe_node_gfs=$(( npe_node_max / nth_gfs ))
```
```suggestion
    wtime="00:30:00"
    npe=$(( layout_x * layout_y * 6 ))
    nth=1
    npe_node=$(( npe_node_max / nth ))
```

DavidHuber-NOAA committed:
Update parm/config/gfs/config.resources

DavidHuber-NOAA commented on parm/config/gfs/config.resources:
```diff
@@ -819,568 +761,523 @@ case ${step} in
+    npe=40
+    nth=8
+    npe_node=10
+    npe_postsndcfp=9
+    npe_node_postsndcfp=1
```
```suggestion
    export npe_postsndcfp=9
    export npe_node_postsndcfp=1
```

DavidHuber-NOAA committed:
Export postsndcfp resource variables

aerorahul commented on env/AWSPW.env:
```diff
@@ -34,81 +45,43 @@ if [[ "${step}" = "fcst" ]] || [[ "${step}" = "efcs" ]]; then
-    export NTHREADS_FIT2OBS=${nth_fit2obs:-1}
-    [[ ${NTHREADS_FIT2OBS} -gt ${nth_max} ]] && export NTHREADS_FIT2OBS=${nth_max}
-    export MPIRUN="${launcher} -n ${npe_fit2obs}"
+    export NTHREADS_FIT2OBS=${NTHREADS1}
```
on the RHS is `NTHREADS1`.  Just checking of that is a typo or meant to be `${NTHREADS:-1}`

aerorahul replied:
Looking further, it seems `NTHREADS1=${nth:-1}`, so this is not a typo

aerorahul commented on env/HERA.env:
```diff
@@ -31,9 +31,20 @@ if [[ -n "${SLURM_JOB_ID:-}" ]]; then
+    nth_max=$((npe_node_max / npe_node))
+    NTHREADSmax=${nth:-${nth_max}}
+    NTHREADS1=${nth:-1}
+    NTHREADS2=${nth:-2}
```
if `nth` is defined, `NTHREADS1` and `NTHREADS2` will be the same.

aerorahul commented on env/HERA.env:
```diff
@@ -31,9 +31,20 @@ if [[ -n "${SLURM_JOB_ID:-}" ]]; then
+    NTHREADS2=${nth:-2}
+    [[ ${NTHREADSmax} -gt ${nth_max} ]] && NTHREADSmax=${nth_max}
+    [[ ${NTHREADS1} -gt ${nth_max} ]] && NTHREADS1=${nth_max}
+    [[ ${NTHREADS2} -gt ${nth_max} ]] && NTHREADS2=${nth_max}
```
What is the purpose of `NTHREADS2`?

DavidHuber-NOAA replied:
`NTHREADS2` is only used by the `awips` job.  I'm not sure why the default needs to be 2 instead of 1 (especially since `config.resources` specifies `nth=1` for the `awips` job), but I kept that way to maintain the current logic.  I'm happy to change it to `NTHREADS1`.

DavidHuber-NOAA replied:
Removed and used `NTHREADS1` instead.

aerorahul commented on jobs/JGDAS_ATMOS_ANALYSIS_DIAG:
```diff
@@ -24,6 +23,7 @@ export gcyc=${GDATE:8:2}
 export GDUMP="gdas"
 export GDUMP_ENS="enkf${GDUMP}"
 
+CDUMP="${RUN/enkf}"
```
Since this is not exported, should this be even declared?
i.e. why not the line 27 be:
```
export OPREFIX="${RUN/enkf}.t${cyc}z."
```

DavidHuber-NOAA replied:
Removed.

aerorahul commented on parm/config/gefs/config.resources:
```diff
@@ -53,38 +53,38 @@ export npe_node_max
-    export npe_stage_ic=1
-    export npe_node_stage_ic=1
-    export nth_stage_ic=1
+    export wtime="00:15:00"
```
can we use `walltime`?

DavidHuber-NOAA replied:
Done

aerorahul commented on parm/config/gefs/config.resources:
```diff
@@ -53,38 +53,38 @@ export npe_node_max
+    export wtime="00:15:00"
+    export npe=1
+    export npe_node=1
+    export nth=1
```
can we use `threads_per_task`?

DavidHuber-NOAA replied:
Done

aerorahul commented on parm/config/gefs/config.resources:
```diff
@@ -53,38 +53,38 @@ export npe_node_max
-    export nth_stage_ic=1
+    export wtime="00:15:00"
+    export npe=1
+    export npe_node=1
```
can we use `tasks_per_node`?

aerorahul commented on parm/config/gefs/config.resources:
```diff
@@ -53,38 +53,38 @@ export npe_node_max
-    export npe_node_stage_ic=1
-    export nth_stage_ic=1
+    export wtime="00:15:00"
+    export npe=1
```
Do we need this? Or call this `tasks`

DavidHuber-NOAA replied:
`tasks` gets ambiguous in the setup scripts given that jobs (e.g. `gdasfcst`) are referred to as `tasks`, which is reinforced by the `wxflow` `Task` class.  Perhaps we could call it `job_tasks` or `tasks_per_job`?

DavidHuber-NOAA replied:
Done (went with ntasks).

aerorahul commented on parm/config/gefs/config.resources:
```diff
@@ -53,38 +53,38 @@ export npe_node_max
+    export npe=12
+    export nth=1
+    export npe_node=$(( npe_node_max / nth ))
+    export NTASKS=${npe}
```
For a later PR, but what is the purpose of redefining the same thing with a different name.

DavidHuber-NOAA replied:
It can probably be removed in favor of `ntasks`.  There are just a few wave jobs that use it for MPMD control.

aerorahul reviewed (commented):
some early comments

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Update ufs model hash to 20240625 (#2729)

DavidHuber-NOAA committed:
Remove additional CDUMP references

DavidHuber-NOAA committed:
Rename resource variables

npe=>ntasks
npe_node=>tasks_per_node
npe_node_max=>max_tasks_per_node
nth=>threads_per_task
wtime=>walltime

DavidHuber-NOAA committed:
Disambiguate calcanl ntasks

DavidHuber-NOAA committed:
Correct WCOSS2 mpiexec call for fcst/efcs jobs

DavidHuber-NOAA committed:
Remove NTHREADS2

DavidHuber-NOAA commented:
Running CI on Hercules.

TerrenceMcGuinness-NOAA commented:
BTW, There is a PR in to "catch" create_expermen.py fails for non-Jenkins reporting.
This CI Test failed on create experiment for **C48_S2SWA_gefs** (only) on Hera:

```
mterry (hercules-login-2) ~ $  /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/scripts/utils/ci_utils_wrapper.sh create_experiment /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml
Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: None
2024-07-08 12:51:27,001 - INFO     - root        : BEGIN: __main__.input_args
2024-07-08 12:51:27,001 - DEBUG    - root        : (  )
2024-07-08 12:51:27,002 - INFO     - root        :   END: __main__.input_args
2024-07-08 12:51:27,002 - DEBUG    - root        :  returning: Namespace(yaml=PosixPath('/work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml'), overwrite=True)
2024-07-08 12:51:27,012 - INFO     - root        : Call: setup_expt.main()
2024-07-08 12:51:27,013 - DEBUG    - root        : setup_expt.py gefs forecast-only --pslot C48_S2SWA_gefs_972cf2b9 --app S2SWA --resdetatmos 48 --resdetocean 5.0 --resensatmos 48 --nens 2 --gfs_cyc 1 --start cold --comroot UNDEFINED/COMROOT --expdir UNDEFINED/EXPDIR --idate 2021032312 --edate 2021032312 --yaml /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/yamls/gefs_ci_defaults.yaml --overwrite

directory already exists in UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9
removing directory ........ UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9


directory already exists in UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
removing directory ........ UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9

****************************************************************************************************
EXPDIR: UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
ROTDIR: UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9
****************************************************************************************************
2024-07-08 12:51:27,022 - INFO     - root        : Call: setup_xml.main()
2024-07-08 12:51:27,022 - DEBUG    - root        : setup_xml.py /work2/noaa/stmp/CI/HERCULES/2727/gefs/UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
Finalizing initialize
sourcing config.stage_ic
sourcing config.fcst
sourcing config.atmos_products
sourcing config.efcs
sourcing config.atmos_ensstat
sourcing config.waveinit
sourcing config.wavepostsbs
sourcing config.wavepostpnt
sourcing config.oceanice_products
sourcing config.prep_emissions
Source configs with RUN=gefs
Traceback (most recent call last):
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs//workflow/create_experiment.py", line 108, in <module>
    setup_xml.main(setup_xml_args)
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/tasks.py", line 245, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/gefs_tasks.py", line 72, in stage_ic
    resources = self.get_resource('stage_ic')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/tasks.py", line 192, in get_resource
    threads = task_config[f'threads_per_task']
              ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/wxflow/attrdict.py", line 84, in __missing__
    raise KeyError(name)
KeyError: 'threads_per_task'
```

TerrenceMcGuinness-NOAA commented:
BTW, There is a PR in to "catch"` create_expermen.py` fails for non-Jenkins reporting.
This CI Test failed on create experiment for **C48_S2SWA_gefs** (only) on Hercules:

```
mterry (hercules-login-2) ~ $  /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/scripts/utils/ci_utils_wrapper.sh create_experiment /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml
Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: None
2024-07-08 12:51:27,001 - INFO     - root        : BEGIN: __main__.input_args
2024-07-08 12:51:27,001 - DEBUG    - root        : (  )
2024-07-08 12:51:27,002 - INFO     - root        :   END: __main__.input_args
2024-07-08 12:51:27,002 - DEBUG    - root        :  returning: Namespace(yaml=PosixPath('/work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/pr/C48_S2SWA_gefs.yaml'), overwrite=True)
2024-07-08 12:51:27,012 - INFO     - root        : Call: setup_expt.main()
2024-07-08 12:51:27,013 - DEBUG    - root        : setup_expt.py gefs forecast-only --pslot C48_S2SWA_gefs_972cf2b9 --app S2SWA --resdetatmos 48 --resdetocean 5.0 --resensatmos 48 --nens 2 --gfs_cyc 1 --start cold --comroot UNDEFINED/COMROOT --expdir UNDEFINED/EXPDIR --idate 2021032312 --edate 2021032312 --yaml /work2/noaa/stmp/CI/HERCULES/2727/gefs/ci/cases/yamls/gefs_ci_defaults.yaml --overwrite

directory already exists in UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9
removing directory ........ UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9


directory already exists in UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
removing directory ........ UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9

****************************************************************************************************
EXPDIR: UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
ROTDIR: UNDEFINED/COMROOT/C48_S2SWA_gefs_972cf2b9
****************************************************************************************************
2024-07-08 12:51:27,022 - INFO     - root        : Call: setup_xml.main()
2024-07-08 12:51:27,022 - DEBUG    - root        : setup_xml.py /work2/noaa/stmp/CI/HERCULES/2727/gefs/UNDEFINED/EXPDIR/C48_S2SWA_gefs_972cf2b9
Finalizing initialize
sourcing config.stage_ic
sourcing config.fcst
sourcing config.atmos_products
sourcing config.efcs
sourcing config.atmos_ensstat
sourcing config.waveinit
sourcing config.wavepostsbs
sourcing config.wavepostpnt
sourcing config.oceanice_products
sourcing config.prep_emissions
Source configs with RUN=gefs
Traceback (most recent call last):
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs//workflow/create_experiment.py", line 108, in <module>
    setup_xml.main(setup_xml_args)
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/tasks.py", line 245, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/gefs_tasks.py", line 72, in stage_ic
    resources = self.get_resource('stage_ic')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/rocoto/tasks.py", line 192, in get_resource
    threads = task_config[f'threads_per_task']
              ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/work2/noaa/stmp/CI/HERCULES/2727/gefs/workflow/wxflow/attrdict.py", line 84, in __missing__
    raise KeyError(name)
KeyError: 'threads_per_task'
```

DavidHuber-NOAA committed:
Rename threads to threads_per_task for a few gefs jobs

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA commented:
Offline GEFS test passed on Hercules after a bug fix.  Rerunning CI.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  CI maintenance updates and adding CI Unit Tests (#2740)
  Adds contents of constructor and initialize methods to marine LETKF class (#2635)
  Fix GDAS group B restart archiving (#2735)
  Add fcst dependency to ocnanalprep (#2728)
  Update (partially) global-workflow for orion+rocky9 (#2741)

DavidHuber-NOAA committed:
Merge branch 'feature/simplify_res' of github.com:davidhuber-noaa/global-workflow into feature/simplify_res

* 'feature/simplify_res' of github.com:davidhuber-noaa/global-workflow:
  Rename threads to threads_per_task for a few gefs jobs

DavidHuber-NOAA committed:
Fix tasks per node variable name

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA A CI unit test is failing to run `create_experiment.py` for the `C48_ATM` test case.  I can run this same command on Hera without issue.  Can you please take a look at this and let me know if there is something I need to fix or if this may be an issue with the CI unit tests?

DavidHuber-NOAA commented:
@InnocentSouopgui-NOAA Would you be able to test this PR on S4 and/or Jet?  It has comprehensive changes for resource calculations and modifies each machine's `env` file accordingly.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Updated prepobs and fit2obs versions for Orion Rocky9 (#2758)

aerorahul reviewed (dismissed):
Changes are extensive and look good.

I reviewed this PR in bits and pieces since this is a large update.
I suggest other reviewers to also review this carefully, as it is possible to have overlooked/missed.

Additionally, this PR should be tested to the fullest on all available platforms as extensively as possible to minimize disruptions to development.

InnocentSouopgui-NOAA commented:
> @InnocentSouopgui-NOAA Would you be able to test this PR on S4 and/or Jet? It has comprehensive changes for resource calculations and modifies each machine's `env` file accordingly.

Sure, will do.

DavidHuber-NOAA committed:
Empty commit to trigger CI unit tests

DavidHuber-NOAA committed:
Revert wxflow and applications.py

DavidHuber-NOAA committed:
Check for NoneType after calling which; rename rocotostat to prevent nameclash with rocotostat module

DavidHuber-NOAA committed:
Update wxflow hash

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

DavidHuber-NOAA committed:
Try unsetting for CI test

DavidHuber-NOAA committed:
Revert unset

DavidHuber-NOAA committed:
Fix tasks/node variable for WCOSS2 fcst/efcs jobs

InnocentSouopgui-NOAA commented:
Successful cycling of resolution 96/48 on S4. Is there a need to test multiple resolution?
I am starting the test on Jet

DavidHuber-NOAA commented:
@InnocentSouopgui-NOAA Yeah, this has changes for S4 analysis and eupd jobs at C384 as well, though I don't know if anyone is running that resolution on S4 these days.  If not, then it can be skipped.

Similarly, Jet has modifications for C384 and C48-C192 (a single test at C96/C48 and another at C384 would suffice).  Lastly, there is a change at C768 for the ecen job on Jet, but I don't think anyone is running that resolution on Jet, so I think it can be skipped.

WalterKolczynski-NOAA commented on env/AWSPW.env:
```diff
@@ -19,96 +19,64 @@ export NTHSTACK=1024000000
+    [[ ${NTHREADSmax} -gt ${max_threads_per_task} ]] && NTHREADSmax=${max_threads_per_task}
+    [[ ${NTHREADS1} -gt ${max_threads_per_task} ]] && NTHREADS1=${max_threads_per_task}
+    APRUN="${launcher} -n ${ntasks}"
+fi
```
What happens if the variables _aren't_ set?

DavidHuber-NOAA replied:
There are a handful of jobs that do not set these variables because they don't source config.resources.  In particular, `gfsgempakncdccupapgif` and `gfsnpoess_pgrb2_0p5deg`.  If the `WAFS` tasks were ever added to the mesh, they would also run into this issue.  Since these only run on WCOSS2, I could add an `elif` block in `WCOSS2.env` to check if one of these jobs is sourcing it or otherwise raise an error.  For the other `env` files, I will just raise an error.

aerorahul replied:
WAFS is definitely (may be) not going to be added back.

WalterKolczynski-NOAA replied:
Those jobs' configs should have sourcing of `config.resources` added like all the other jobs, with blocks added to the env files as well if needed.

DavidHuber-NOAA replied:
The issue was that the job configs were not sourced in the JJobs.  I've added that now and the `.env` files will raise errors if `config.resources` is not sourced first.

WalterKolczynski-NOAA commented on env/HERA.env:
```diff
@@ -157,167 +140,128 @@ elif [[ "${step}" = "anal" ]] || [[ "${step}" = "analcalc" ]]; then
-    [[ -z "${!ppn+0}" ]] && ppn="npe_node_${step}"
-    nprocs="npe_${step}_${RUN}"
-    [[ -z ${!nprocs+0} ]] && nprocs="npe_${step}"
+    ppn="tasks_per_node"
+    [[ -z "${!ppn+0}" ]] && ppn="tasks_per_node"
+    nprocs="ntasks"
+    [[ -z ${!nprocs+0} ]] && nprocs="ntasks"
```
Setting the variable again to the same thing isn't going to fix an issue.
```suggestion
    ppn="tasks_per_node"
    nprocs="ntasks"
```

WalterKolczynski-NOAA replied:
Also, why is the formulation for Hera and WCOSS different from other machines?

DavidHuber-NOAA replied:
`ppn` and `nprocs` are not needed anymore.  I will delete these from `HERA.env` and `WCOSS2.env`.

DavidHuber-NOAA replied:
Fixed.

WalterKolczynski-NOAA commented on jobs/JGDAS_ATMOS_ANALYSIS_DIAG:
```diff
@@ -24,7 +23,7 @@ export gcyc=${GDATE:8:2}
 export GDUMP="gdas"
```
If we are eliminating this type of use for `CDUMP`, it stands to reason `GDUMP` should also be changed to `GRUN` (or, even better, `RUN_GUESS` or similar).

WalterKolczynski-NOAA replied:
(This can be done in a subsequent PR.)

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_ANALYSIS:
```diff
@@ -8,7 +8,7 @@ source "${HOMEgfs}/ush/jjob_header.sh" -e "anal" -c "base anal"
 ##############################################
 export CDATE=${CDATE:-${PDY}${cyc}}
-export CDUMP=${RUN/enkf}
+export rCDUMP=${RUN/enkf}
```
Same argument as `GDUMP`

WalterKolczynski-NOAA commented on jobs/rocoto/prep.sh:
```diff
@@ -13,7 +13,7 @@ export job="prep"
 source "${HOMEgfs}/ush/jjob_header.sh" -e "prep" -c "base prep"
 
-export CDUMP="${RUN/enkf}"
+CDUMP="${RUN/enkf}"
```
Add a comment about downstream script using `CDUMP`?

DavidHuber-NOAA replied:
It's not in this case.  I have renamed this variable to `RUN_local` to make it more clear and added a comment.

WalterKolczynski-NOAA commented on parm/config/gefs/config.resources:
```diff
@@ -16,75 +16,75 @@ step=$1
     export is_exclusive=True
 
-    _CDUMP_LIST=${CDUMP:-"gdas gfs"}
+    _RUN_LIST=${RUN:-"gefs gfs"}
```
GEFS only has the one `RUN`, so I think all these can be eliminated and GEFS just always uses the `_gfs` version.

DavidHuber-NOAA replied:
Fixed.

WalterKolczynski-NOAA commented on parm/config/gfs/config.anal:
```diff
@@ -12,16 +12,13 @@ if [[ ${DONST} = "YES" ]]; then
 fi
 
-if [[ "${CDUMP}" = "gfs" ]] ; then
+if [[ "${RUN}" = "gfs" ]] ; then
```
```suggestion
if [[ "${RUN}" == "gfs" ]] ; then
```

WalterKolczynski-NOAA commented on parm/config/gfs/config.eobs:
```diff
@@ -11,12 +11,11 @@ echo "BEGIN: config.eobs"
 export OBSINPUT_INVOBS="dmesh(1)=225.0,dmesh(2)=225.0,dmesh(3)=225.0,dmesh(4)=100.0"
 export OBSQC_INVOBS="tcp_width=60.0,tcp_ermin=2.0,tcp_ermax=12.0"
-if [ $LEVS = "128" ]; then
+if [ ${LEVS} = "128" ]; then
```
```suggestion
if (( LEVS == 128 )); then
```

DavidHuber-NOAA committed:
Increase fit2obs memory at high res

WalterKolczynski-NOAA commented on ush/calcanl_gfs.py:
```diff
@@ -19,7 +19,7 @@
                 ComIn_Ges, GPrefix,
                 FixDir, atmges_ens_mean, RunDir, NThreads, NEMSGet, IAUHrs,
-                ExecCMD, ExecCMDMPI, ExecAnl, ExecChgresInc, Cdump, JEDI):
+                ExecCMD, ExecCMDMPI, ExecAnl, ExecChgresInc, Run, JEDI):
```
All of these should be in snake case since they are variables and not classes, but for this PR:
```suggestion
                ExecCMD, ExecCMDMPI, ExecAnl, ExecChgresInc, run, JEDI):
```

WalterKolczynski-NOAA commented on workflow/applications/applications.py:
```diff
@@ -31,7 +32,11 @@ def __init__(self, conf: Configuration) -> None:
         self.scheduler = Host().scheduler
 
-        _base = conf.parse_config('config.base')
+        # Save the configuration so we can source the config files when
+        # determining task resources
+        self.conf = conf
```
Instead of doing this and then re-parsing all the config files in Task initialization, how about using the first-pass of `_base` to construct a list of runs, and then creating a dictionary that gets used for all tasks, changing `self.configs[task]` to `self.configs[run][task]`.

WalterKolczynski-NOAA replied:
I think doing it this way will also allow us to keep some methods as private you are currently changing to public.

DavidHuber-NOAA replied:
Opened issue #2783 to tackle this at a later time.

WalterKolczynski-NOAA commented on workflow/applications/applications.py:
```diff
@@ -31,7 +32,11 @@ def __init__(self, conf: Configuration) -> None:
+        _base = self.conf.parse_config('config.base')
         # Define here so the child __init__ functions can use it; will
         # be overwritten later during _init_finalize().
         self._base = _base
```
I think we want to get rid of this in the future, since is should be the same as `self.configs[run]['base']`, but that will require a lot of knock-on changes and we don't want to complicate this PR further.

DavidHuber-NOAA replied:
Will tackle in #2783

WalterKolczynski-NOAA commented on parm/config/gfs/config.resources:
```diff
@@ -440,273 +408,242 @@ case ${step} in
 
     if [[ "${step}" == "fcst" ]]; then
-      _CDUMP_LIST=${CDUMP:-"gdas gfs"}
+      _RUN_LIST=${RUN:-"gdas gfs"}
     elif [[ "${step}" == "efcs" ]]; then
-      _CDUMP_LIST=${CDUMP:-"enkfgdas enkfgfs"}
+      _RUN_LIST=${RUN:-"enkfgdas enkfgfs"}
     fi
```
If the rocoto has a different config for each `RUN`, I don't think we need any of this `RUN_LIST` stuff anymore.

DavidHuber-NOAA replied:
I think you're right.  I'll test this out.

DavidHuber-NOAA replied:
Fixed.

WalterKolczynski-NOAA commented on parm/config/gfs/config.resources:
```diff
@@ -819,568 +756,524 @@ case ${step} in
-      declare -x "npe_${step}_${_CDUMP}"="${NTASKS_TOT}"
-      declare -x "nth_${step}_${_CDUMP}"="${UFS_THREADS}"
-      declare -x "npe_node_${step}_${_CDUMP}"="${npe_node_max}"
+      declare -x "ntasks_${_RUN}"="${NTASKS_TOT}"
+      declare -x "threads_per_task_${_RUN}"="${UFS_THREADS}"
+      declare -x "tasks_per_node_${_RUN}"="${max_tasks_per_node}"
```
Similarly, I don't think we need to tag these with a `_RUN` anymore. Or the wallclock below.

DavidHuber-NOAA replied:
Fixed

WalterKolczynski-NOAA commented on parm/config/gfs/config.resources.HERA:
```diff
@@ -5,30 +5,30 @@
-      export npe_anal_gfs=270
-      export nth_anal=8
-      export npe_node_anal=$(( npe_node_max / nth_anal ))
+      export ntasks_gdas=270
+      export ntasks_gfs=270
```
```suggestion
      export ntasks=270
```

WalterKolczynski-NOAA commented on parm/config/gfs/config.resources.JET:
```diff
@@ -5,10 +5,10 @@
-      export npe_anal_gfs=270
-      export nth_anal=8
-      export npe_node_anal=$(( npe_node_max / nth_anal ))
+      export ntasks_gdas=270
+      export ntasks_gfs=270
```
```suggestion
      export ntasks=270
```

WalterKolczynski-NOAA reviewed (changes requested):
I think @aerorahul is going to schedule a meeting with the three of us to discuss some of the bigger things.

InnocentSouopgui-NOAA commented:
Successful cycling of resolution 96/48 on Jet.
Starting 384.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Set METplus process count in config.metp; add verif-global support for Rocky 9 (#2774)
  Hotfix: Update jcb to avoid git-lfs files (#2782)

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Replace Jinja namespaces with replace_tmpl filter and disable ACCOUNT_SERVICE option (#2775)
  Add task to process reforecast variables to save on WCOSS2 (#2680)

DavidHuber-NOAA committed:
Remove reference to account_service

DavidHuber-NOAA committed:
Remove unused ppn and nprocs from env files

DavidHuber-NOAA committed:
Require config.resources to be sourced for each job

DavidHuber-NOAA committed:
Replace CDUMP with RUN_local/RUN

DavidHuber-NOAA committed:
Remove 'gefs' from default _RUN_LIST

DavidHuber-NOAA committed:
Apply suggestions from code review

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

DavidHuber-NOAA committed:
Remove unneeded _RUN_LIST

DavidHuber-NOAA committed:
Merge branch 'feature/simplify_res' of github.com:DavidHuber-NOAA/global-workflow into feature/simplify_res

DavidHuber-NOAA committed:
Sync with develop and resolve conflicts

DavidHuber-NOAA committed:
Remove local declaration

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

aerorahul reviewed (dismissed):
looks good.

When this PR is merged, an announcement will be appreciated.  
I think there will be lots of developers that will get conflicts; especially those adding new jobs or tuning resources, so we should be ready to assist.

DavidHuber-NOAA committed:
Merge branch 'feature/simplify_res' of github.com:davidhuber-noaa/global-workflow into feature/simplify_res

* 'feature/simplify_res' of github.com:davidhuber-noaa/global-workflow:
  Temporarily disable METplus jobs (#2796)
  Remove local declaration
  Remove unneeded _RUN_LIST
  Apply suggestions from code review
  Remove 'gefs' from default _RUN_LIST
  Replace CDUMP with RUN_local/RUN
  Require config.resources to be sourced for each job
  Remove unused ppn and nprocs from env files
  Refactoring of the marine B-matrix job (#2749)

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  HR4 GWD update (#2732)

aerorahul reviewed (dismissed):
looks good.  Approving again ..

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/simplify_res

* emc/develop:
  Remove f000 from atmos rocoto tasks for replay cases (#2778)
